### PR TITLE
Make blocks to be keyed on UnixNano instead of time.Time

### PR DESCRIPTION
**Pull request by cw9**
To void map miss caused by different locale set in time.Time

@xichen2020 @robskillington

### CODE REVIEW

File: storage/block/block.go [hunk 1]
```diff
@@ -36,7 +37,10 @@ var (
 	errReadFromClosedBlock         = errors.New("attempt to read from a closed block")
 	errRetrievableBlockNoRetriever = errors.New("attempt to read from a retrievable block with no retriever")
 
-	timeZero = time.Time{}
+	timeUnixNanoZero = int64(math.MinInt64)
+	// NB(cw): the behavior of calling UnixNano() on time.Time{} is undefined
+	// using this timeZero makes testing a bit easier.
+	timeZero = time.Unix(0, timeUnixNanoZero)
```
Comment (xichen2020) [inline storage/block/block.go]: Why not just `time.Unix(0, 0)`

File: storage/block/block.go [hunk 2]
```diff
@@ -326,40 +332,45 @@ func (dbb *databaseSeriesBlocks) AddSeries(other DatabaseSeriesBlocks) {
...
+	return time.Unix(0, dbb.max)
 }
 
 func (dbb *databaseSeriesBlocks) BlockAt(t time.Time) (DatabaseBlock, bool) {
-	b, ok := dbb.elems[t]
+	b, ok := dbb.elems[t.UnixNano()]
 	return b, ok
 }
 
 func (dbb *databaseSeriesBlocks) AllBlocks() map[time.Time]DatabaseBlock {
```
Comment (xichen2020) [inline storage/block/block.go]: hm..where do we use this function?

File: storage/block/block.go (hunk 1 above)
Comment (robskillington) [inline storage/block/block.go]: Or even just `timeZero = int64(0)`

File: storage/block/block.go [hunk 3]
```diff
@@ -326,40 +332,45 @@ func (dbb *databaseSeriesBlocks) AddSeries(other DatabaseSeriesBlocks) {
...
+	res := make(map[time.Time]DatabaseBlock, len(dbb.elems))
+	for nano, db := range dbb.elems {
+		res[time.Unix(0, nano)] = db
+	}
+	return res
 }
 
 func (dbb *databaseSeriesBlocks) RemoveBlockAt(t time.Time) {
-	if _, exists := dbb.elems[t]; !exists {
+	tNano := t.UnixNano()
```
Comment (robskillington) [inline storage/block/block.go]: Maybe a better name is `unixNano := t.UnixNano()`?

File: storage/block/block.go [hunk 4]
```diff
@@ -326,40 +332,45 @@ func (dbb *databaseSeriesBlocks) AddSeries(other DatabaseSeriesBlocks) {
...
 
 func (dbb *databaseSeriesBlocks) BlockAt(t time.Time) (DatabaseBlock, bool) {
-	b, ok := dbb.elems[t]
+	b, ok := dbb.elems[t.UnixNano()]
 	return b, ok
 }
 
 func (dbb *databaseSeriesBlocks) AllBlocks() map[time.Time]DatabaseBlock {
-	return dbb.elems
+	res := make(map[time.Time]DatabaseBlock, len(dbb.elems))
```
Comment (robskillington) [inline storage/block/block.go]: Yeah we can't create a new map each time, this is called on some huge maps. A copy each time is extremely expensive (think millions of IDs all getting `AllBlocks(...)` called on it.

**Comment by robskillington** [comment]
Need to also update the following places that `map[time.Time]...` is used:

```
persist/fs/seek_manager.go (2 definitions)
storage/bootstrap/bootstrapper/commitlog/source.go (1 definition)
storage/repair.go (1 definition)
storage/repair/metadata.go (1 definition)
storage/repair/types.go (1 definition)
storage/shard.go (1 definition)
```

File: storage/block/block.go (hunk 1 above)
Comment (cw9) [inline storage/block/block.go]: I kinda want to make timeUnixNanoZero and timeZero interchangeable, gonna do:
	timeZero         = time.Unix(0, 0)
	timeUnixNanoZero = timeZero.UnixNano()

File: storage/block/block.go (hunk 3 above)
Comment (cw9) [inline storage/block/block.go]: sure

File: storage/block/block.go (hunk 4 above)
Comment (cw9) [inline storage/block/block.go]: Sure, then I'll change the interface to return map[int64]DatabaseBlock

**Comment by cw9** [comment]
Yeah, I'm thinking to keep the blast radiance down per diff, I'm gonna fix all the usage of this interface in this one and handle those other places in separate diffs.

**Comment by cw9** [comment]
Talked with Rob, will postpone this pr and do a full sweep of time.Time in maps across the code base

**Pull request closed without merging**